src/ripple/shamap/SHAMapItem.h

```diff
@@ -20,52 +20,189 @@
+    // arbitrary construction.
+    SHAMapItem(uint256 const& tag, Slice data)
+        : tag_(tag), size_(static_cast<std::uint32_t>(data.size()))
     {
+        std::memcpy(
+            reinterpret_cast<std::uint8_t*>(this) + sizeof(*this),
+            data.data(),
+            data.size());
     }
 
+public:
+    SHAMapItem() = delete;
+
+    SHAMapItem(SHAMapItem const& other) = delete;
+
+    SHAMapItem&
+    operator=(SHAMapItem const& other) = delete;
+
+    SHAMapItem(SHAMapItem&& other) = delete;
+
+    SHAMapItem&
+    operator=(SHAMapItem&&) = delete;
+
     uint256 const&
     key() const
     {
         return tag_;
     }
 
-    Slice
-    slice() const
-    {
-        return static_cast<Slice>(data_);
-    }
-
     std::size_t
     size() const
     {
-        return data_.size();
+        return size_;
     }
 
     void const*
     data() const
     {
-        return data_.data();
+        return reinterpret_cast<std::uint8_t const*>(this) + sizeof(*this);
+    }
+
+    Slice
+    slice() const
+    {
+        return {data(), size()};
     }
 };
 
+namespace detail {
+
+// clang-format off
+inline SlabAllocator<SHAMapItem,  128> slab128 ( 7000000);
+inline SlabAllocator<SHAMapItem,  192> slab192 ( 1000000);
+inline SlabAllocator<SHAMapItem,  256> slab256 (10000000);
+inline SlabAllocator<SHAMapItem,  512> slab512 (  750000);
+inline SlabAllocator<SHAMapItem, 1024> slab1024(  450000);
+// clang-format on
+
+inline std::atomic<std::uint64_t> cnt64 = 0;
+
+}  // namespace detail
+
+inline void
+intrusive_ptr_add_ref(SHAMapItem const* x)
+{
+    // This can only happen if someone releases the last reference to the
+    // item while we were trying to increment the refcount.
+    if (x->refcount_++ == 0)
+        LogicError("SHAMapItem: the reference count is 0!");
+}
+
+inline void
+intrusive_ptr_release(SHAMapItem const* x)
+{
+    if (--x->refcount_ == 0)
+    {
+        auto p = reinterpret_cast<std::uint8_t const*>(x);
+
+        // The SHAMapItem constuctor isn't trivial (because the destructor
+        // for CountedObject isn't) so we can't avoid calling it here, but
+        // plan for a future where we might not need to.
+        if constexpr (!std::is_trivially_destructible_v<SHAMapItem>)
+            std::destroy_at(x);
+
+        // At most one slab will claim this pointer; if none do, it was
+        // allocated manually, so we free it manually.
+        if (!detail::slab128.free(p) && !detail::slab192.free(p) &&
+            !detail::slab256.free(p) && !detail::slab512.free(p) &&
+            !detail::slab1024.free(p))
+        {
+            std::free(const_cast<std::uint8_t*>(p));
+        }
+    }
+}
+
+inline boost::intrusive_ptr<SHAMapItem>
+make_shamapitem(uint256 const& tag, Slice data)
+{
+    assert(data.size() <= megabytes<std::size_t>(64));
+
+    std::uint8_t* raw = [&data]() -> std::uint8_t* {
+        if (data.size() < 128)
```

Comment:
Should these be <= instead of < ? If not, what's the extra byte in the slab for?